ambry-store/src/main/java/com.github.ambry.store/Log.java

```diff
@@ -364,23 +390,82 @@ private void rollOverIfRequired(long writeSize) throws IOException {
     long segmentCapacity = activeSegment.getCapacityInBytes();
     if (writeSize > segmentCapacity - LogSegment.HEADER_SIZE) {
       metrics.overflowWriteError.inc();
       throw new IllegalArgumentException("Write of size [" + writeSize + "] cannot be serviced because it is greater "
           + "than a single segment's capacity [" + (segmentCapacity - LogSegment.HEADER_SIZE) + "]");
     }
-    String newSegmentName = LogSegmentNameHelper.getNextPositionName(activeSegment.getName());
-    logger.info("Allocating new segment with name: " + newSegmentName);
-    File newSegmentFile = allocate(LogSegmentNameHelper.nameToFilename(newSegmentName), segmentCapacity);
-    LogSegment newSegment = new LogSegment(newSegmentName, newSegmentFile, segmentCapacity, metrics, true);
-    segmentsByName.put(newSegmentName, newSegment);
+    if (remainingUnallocatedSegments.decrementAndGet() < 0) {
+      remainingUnallocatedSegments.incrementAndGet();
+      metrics.overflowWriteError.inc();
+      throw new IllegalStateException(
+          "There is no more capacity left in [" + dataDir + "]. Max capacity is [" + capacityInBytes + "]");
+    }
+    Pair<String, String> segmentNameAndFilename = getNextSegmentNameAndFilename();
+    logger.info("Allocating new segment with name: " + segmentNameAndFilename.getFirst());
+    File newSegmentFile = allocate(segmentNameAndFilename.getSecond(), segmentCapacity);
+    LogSegment newSegment =
+        new LogSegment(segmentNameAndFilename.getFirst(), newSegmentFile, segmentCapacity, metrics, true);
+    segmentsByName.put(segmentNameAndFilename.getFirst(), newSegment);
+  }
+
+  /**
+   * @return the name and filename of the segment that is to be created.
+   */
+  private Pair<String, String> getNextSegmentNameAndFilename() {
+    Pair<String, String> nameAndFilename;
+    if (segmentNameAndFileNameIterator != null && segmentNameAndFileNameIterator.hasNext()) {
+      nameAndFilename = segmentNameAndFileNameIterator.next();
+    } else if (activeSegment == null) {
+      // this code path gets exercised only on first startup
+      String name = LogSegmentNameHelper.generateFirstSegmentName(isLogSegmented);
+      nameAndFilename = new Pair<>(name, LogSegmentNameHelper.nameToFilename(name));
+    } else {
+      String name = LogSegmentNameHelper.getNextPositionName(activeSegment.getName());
+      nameAndFilename = new Pair<>(name, LogSegmentNameHelper.nameToFilename(name));
+    }
+    return nameAndFilename;
+  }
+
+  /**
+   * Adds a {@link LogSegment} instance to the log.
+   * @param segment the {@link LogSegment} instance to add.
+   * @param increaseUsedSegmentCount {@code true} if the number of segments used has to be incremented, {@code false}
+   *                                             otherwise.
+   * @throws IllegalArgumentException if the {@code segment} being added is past the active segment.
+   */
+  void addSegment(LogSegment segment, boolean increaseUsedSegmentCount) {
```

Comment:
It would be `false` when the caller needs to add segments that are the result of compaction which do not contribute to the allocated segment count (at the time of addition). They will eventually be accounted for when `dropSegment()` calls are made.